lib/grell/crawler.rb

```diff
@@ -51,12 +51,21 @@ def crawl(site, block)
       Grell.logger.info "Visiting #{site.url}, visited_links: #{@collection.visited_pages.size}, discovered #{@collection.discovered_pages.size}"
       site.navigate
       filter!(site.links)
-
-      if block #The user of this block can send us a :retry to retry accessing the page
-        while block.call(site) == :retry
-          Grell.logger.info "Retrying our visit to #{site.url}"
-          site.navigate
-          filter!(site.links)
+      add_redirect_url(site)
+
+      if block # The user of this block can send us a :retry to retry accessing the page
+        begin
+          while block.call(site) == :retry
+            Grell.logger.info "Retrying our visit to #{site.url}"
+            site.navigate
+            filter!(site.links)
+            add_redirect_url(site)
```

Comment:
I suppose it would depend on the how the `add_match_block` is implemented. For the default case, the `add(page)` method would be called twice, but the URL would not be added the second time because it would be filtered out